oeps/oep-0022-bp-django-caches.rst

```diff
@@ -0,0 +1,229 @@
+
+1. There is a common caching bug while using Django caches where someone
+   caches a value that is Falsey (e.g., None or []) and it is mistakenly
+   treated as a cache miss.
+
+2. In Django we commonly use the Memcached back-end. There are various
+   places in code where we read the same value many times within a
+   the scope of a request, and each of these is a network call to
+   Memcached.
+
+3. For testing and business purposes, it is sometimes useful to be able
+   to invalidate a cached value on a per-request basis.
+
+4. While refactoring caching code and reviewing potential libraries, there
+   seem to be potential best practices that would hide the boilerplate
+   caching logic and better highlight the business logic.
+
+5. There is potential for better standardizing how we handle cache keys.
+   For example, to generate keys that work with Memcached and avoid key
+   collisions in a global space like Memcached.
+
+Note: When and if caching is warranted in any particular circumstance is beyond
+the scope of this OEP.
+
+Decision
+--------
+
+Common Caching Defect and Fix
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following is a common defect when caching using Django caches::
+
+    cached_value = cache.get(key)
+    if not cached_value:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # be treated as a cache miss above.
+    return cached_value
+
+The simplest fix for this defect is::
+
+    _CACHE_MISS = object()
+
+    cached_value = cache.get(key, _CACHE_MISS)
+    if cached_value is _CACHE_MISS:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # no longer be treated as a cache miss above.
+    return cached_value
+
+Use libraries that makes it simpler to avoid the above defect. See
+`Implementations`_ for more details.
+
+Tiered Cache with Request Caching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use a library that allows tiered-caching calls that use a Request Cache backed
+by Memcached. Multiple calls for the same value within a request will not
+require additional network calls to Memcached.
+
+See `Implementations`_ for more details.
+
+Force Cache Refresh
+~~~~~~~~~~~~~~~~~~~
+
+Ensure that new caching functionality has a simple method by which to force a
+cache refresh. This is useful both for testing purposes, as well as for
+business purposes if a problem occurs in Production that requires a capability
+for refreshing a cached value.
+
+Cache Decorators
+~~~~~~~~~~~~~~~~
+
+The following recommendation requires additional assessment.
+
+When caching, create cache methods where the boilerplate caching code is hidden
+behind a decorator. This practice allows the more important business logic to
+stand out.
+
+Note that the boilerplate logic hidden behind the decorator must include
+opinionated code regarding how keys are defined and when and if the key will
```

Comment:
It may not be necessary for the decorator to be opinionated about this. Rather, it can take a function as a parameter to compute the key.

Reply:
1. I corrected the language here.
2. I added an intro to this subsection to clarify its status.
3. I renamed "Decisions" to "Recommendations", and added an intro to that section.
Note: I'm using past tense, but all changes are local for now.  Coming soon.